### PR TITLE
IGNITE-22988 Sql. Align implementations of average calculation (AVG)

### PR DESCRIPTION
**Pull request by AMashenkov**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItAggregatesTest.java [hunk 1]
```diff
@@ -617,14 +613,15 @@ public void testAvgRandom() {
...
+            sql(query, i, NumberUtil.rescaleBigDecimal(num, 0).intValue(), num);
         }
 
         BigDecimal avg = numbers.stream()
                 .reduce(new BigDecimal("0.00"), BigDecimal::add)
-                .divide(BigDecimal.valueOf(numbers.size()), MathContext.DECIMAL64);
+                .divide(BigDecimal.valueOf(numbers.size()), 2, RoundingMode.HALF_UP);
 
         for (String[] rules : makePermutations(DISABLED_RULES)) {
+            System.out.println(String.join(",", rules));
```
Comment (lowka) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItAggregatesTest.java:624]: I think this `system.out.println` is unnecessary.

Comment (lowka) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItAggregatesTest.java:624]: When I got a the following numbers by the random generator:  `{37, 46, 72, 33, 33, 57, 76, 27, 100, 4, 62, 77, 74, 49, 31, 16, 56, 34, 82};`

A got a test failure:

```
Expected: 50 <class java.lang.Integer>
Actual:   51 <class java.lang.Integer>
```
Query results:
```
[[51, 50.84]]
```

Looks like using avg.intValue() ` .returns(avg.intValue(), avg)` is no longer a valid check?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/agg/Accumulators.java [hunk 2]
```diff
@@ -102,13 +103,17 @@ private Supplier<Accumulator> accumulatorFunctionFactory(AggregateCall call) {
...
     private Supplier<Accumulator> avgFactory(AggregateCall call) {
         switch (call.type.getSqlTypeName()) {
+            case TINYINT:
+            case SMALLINT:
+            case INTEGER:
             case BIGINT:
+                return () -> DecimalAvg.FACTORY.apply(0);
             case DECIMAL:
-                return DecimalAvg.FACTORY;
+                // TODO: support intervals.
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/agg/Accumulators.java:112]: let's add link to ticket - IGNITE-15200

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/agg/MapReduceAggregates.java [hunk 3]
```diff
@@ -645,8 +645,15 @@ private static MapReduceAgg createAvgAgg(
 
                 sumDivCnt = rexBuilder.makeCall(IgniteSqlOperatorTable.DECIMAL_DIVIDE, numeratorRef, denominatorRef, p, s);
             } else {
-                RexNode divideRef = rexBuilder.makeCall(SqlStdOperatorTable.DIVIDE, numeratorRef, denominatorRef);
-                sumDivCnt = rexBuilder.makeCast(call.getType(), divideRef, true, false);
+                RelDataType resultType = typeFactory.decimalOf(call.type);
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/agg/MapReduceAggregates.java:648]: I propose to use single code for both cases without any branches. Also use ``false`` for ``matchNullability`` in cast - it needs only for literals, but it's not literals in the case.

**Review by ygerzhedovich — COMMENTED** [review]
Seems we still have not matched types in planne and in execution.
Potentially need to override deriveAvgAggType in our IgniteTypeSystem

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItAggregatesTest.java [hunk 4]
```diff
@@ -23,13 +23,14 @@
...
 import java.math.BigDecimal;
-import java.math.MathContext;
+import java.math.RoundingMode;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Locale;
 import java.util.Random;
 import java.util.stream.Stream;
+import org.apache.calcite.util.NumberUtil;
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItAggregatesTest.java:33]: It's not so good use Calcite utils in our tests

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/agg/MapReduceAggregates.java [hunk 5]
```diff
@@ -631,22 +631,21 @@ private static MapReduceAgg createAvgAgg(
...
-            if (call.getType().getSqlTypeName() == SqlTypeName.DECIMAL) {
-                // Return correct decimal type with correct scale and precision.
-                int precision = call.getType().getPrecision();
-                int scale = call.getType().getScale();
+            RelDataType resultType = typeFactory.decimalOf(call.type);
 
-                RexLiteral p = rexBuilder.makeExactLiteral(BigDecimal.valueOf(precision), tf.createSqlType(SqlTypeName.INTEGER));
-                RexLiteral s = rexBuilder.makeExactLiteral(BigDecimal.valueOf(scale), tf.createSqlType(SqlTypeName.INTEGER));
+            // Return correct decimal type with correct scale and precision.
+            int precision = resultType.getPrecision(); // not used.
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/agg/MapReduceAggregates.java:639]: What reason to have precision at all in this case?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java [hunk 6]
```diff
@@ -361,7 +360,7 @@ public static BigDecimal toBigDecimal(Number value, int precision, int scale) {
      * (see {@link IgniteSqlOperatorTable#DECIMAL_DIVIDE}, their values are ignored at runtime.
      */
     public static BigDecimal decimalDivide(BigDecimal sum, BigDecimal cnt, int p, int s) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:362]: the same question, if precision doesn't used why we have it here? I propose to get rid it

Comment (lowka) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:362]: You can not do this without losing inference of precision/scale (See `DECIMAL_DIVIDE` operator in `IgniteSqlOperatorTable`). Since you can not have different numbers of parameters at the validation phase and in runtime.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/agg/MapReduceAggregates.java (hunk 5 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/agg/MapReduceAggregates.java:639]: seems it ok, based on answer below

**Review by ygerzhedovich — APPROVED** [review]
